Makefile

```diff
@@ -8,7 +8,7 @@ NM ?= $(patsubst %clang,%llvm-nm,$(filter-out ccache sccache,$(CC)))
 ifeq ($(origin AR), default)
 AR = $(patsubst %clang,%llvm-ar,$(filter-out ccache sccache,$(CC)))
 endif
-EXTRA_CFLAGS ?= -O2 -DNDEBUG
+EXTRA_CFLAGS ?= -O2 -DNDEBUG -fPIC -fvisibility=default
```

Comment:
The `-fPIC` overhead also applies to the original non-shared `libc.a` which is not ideal IMHO.

Reply:
Yeah, that's a topic that's come up a few times.  @sunfishcode took a look at libc.a before and after `-fPIC` was added and concluded that it wasn't a big deal from his perspective.  The alternative is to build everything twice -- once for static and once for dynamic.  Either way is fine with me, as long as we have consensus.

Reply:
Are there any disadvantages of building everything twice other than doubled build time? If not, I think that'd be my preference

Reply:
Not that I'm aware of.  @sunfishcode or @alexcrichton do you have any concerns about building the libc and libc++ code twice (with and without `-fPIC`)?  (Actually four times, since we're already building twice for threads/no-threads)

Reply:
I don't have strong opinions either way. The -fPIC overhead from the builds I've seen was low. But also, I don't think building twice is a problem either, unless it makes something more complex.

Reply:
> The -fPIC overhead from the builds I've seen was low

imho it can only be low after a `wasm-opt` run that replaces all symbols with i32 literals again. it would really be nice if the static linking use case doesn't pay for what it doesn't use.

Reply:
I just pushed an update which builds a separate set of `-fPIC` object files for libc.so in addition to the non-`-fPIC` ones for libc.a.

Reply:
I apologize for not reviewing this earlier I've only gotten around to this just now! 

One point I would say in favor of only having one build is that by having two sysroots everyone is now presented with a choice of which one to use and it's possible to get it wrong. That can add downstream complexity to builds, configuration, etc. Having only one build solves this because you can't make the wrong choice and there's only one canonical version of libc.a
 
I see now though I'm late to the punch and you've already updated this to build two copies. I don't feel particularly strongly either way, but I think that this is something to watch out for when integrating this into other toolchains. If it's onerous/difficult to have multiple copies of libc then I think it's worth reconsidering. One example of this is that to use this from Rust there will have to be new target triples, and no preexisting target I think differentiates target triples purely on dynamic linking of the same C standard library.

Reply:
> because you can't make the wrong choice

maybe it is the compiler driver's responsibility to automatically pick between `.a` and `.so` from the same sysroot libdir, based on whether it is doing dynamic linking or not?

Reply:
While possible, that's forcing all compiler drivers, embedders, etc, to have duplicate copies and sysroots. For example the rust standard library would need to be compiled for both dynamic linking and not and store both sysroots, it's not purely a concern for WASI. Rust doesn't currently do this for any other target that I'm aware of, and if that's the case it's unlikely to happen for just the WASI target.

To me at least this cost is not worth the minor savings of code size in libc which optimizing compilers/engines will all strip away anyway.

Reply:
apologies in advance given my rust experience is limited and error-prone; afaik `rust-std` already ships both `libstd-*.so` / `libstd-*.rlib` for various targets, so it looks like a natural conclusion to ship both `-fPIC`/non-`-fPIC` variants of `libstd` for wasm, and for `wasm32-wasi`, ship two variants of `libc` as well (in the same sysroot libdir, to be chosen at link-time)

Reply:
Indeed! Internally though those two artifacts have the same machine code. The standard library is compiled once simultaneously with a dylib/rlib output which means that LLVM generates one copy of all the code for the standard library and then that object file is linked into two locations. This is enabled because all Rust platforms (AFAIK) generate position-independent code by default and eschew the minor size-related optimizations from generating static code. That's left for users to build their own toolchains if that size difference matters.

I'll also admit though I'm probably quite biased on this as I wrote the support for rlibs/dylibs and am the reason that the standard library ships two artifacts for example.